File: Faster Than Recursive/StepFasterThanRecursive.java
import java.util.ArrayList;
import java.lang.StringBuilder;

public class StepFasterThanRecursive
{
  public static void main(String[] args)
  {
    int[] stepList = {1,2,3};
    int n = 40;

    System.out.println(takeSteps(n, stepList));
  }

  static long takeSteps(int n, int[] stepList)
  {
    // the number of ways to climb the stairs is what we eventually want to return
    long nOfWays = 0;
    // will use the number of posible step option a bit
    int nOfStepOptions = stepList.length;

    // first want to gather the maximum posible of each step option from stepOptionList
    int[] maxOfEachStep = new int[nOfStepOptions];

    // maxStop will be used to detect once we have reached the end
    int maxStop = 0;

    // capture maxOfEachStep
    for(int i = 0; i<nOfStepOptions; i++)
    {
      maxOfEachStep[i] = calcMaxOfParStep(n,stepList[i]);
      maxStop = maxStop + (maxOfEachStep[i] * stepList[i]);
    }

    // generate posible combinations of steps of each kind of steps
    // that will get us to n
    int[] potentialCombination = {};
    ArrayList<int[]> stepCombinationList = new ArrayList<int[]>();
    while (true)
    {
      potentialCombination = genNewCombination(potentialCombination,maxOfEachStep);

      if(testStepCombination(potentialCombination,stepList)==n)
      {
        stepCombinationList.add(potentialCombination.clone());
      }
      if(testStepCombination(potentialCombination,stepList)==maxStop)
      {
        break;
      }
    }

    // add up the nOfWays :-)
    for (int i = 0; i<stepCombinationList.size();i++)
    {
      nOfWays = nOfWays + multCombination(stepCombinationList.get(i));
    }

    return nOfWays;
  }

  // fucntion to calculate the maximum number of each aprticular step
  static int calcMaxOfParStep(int n, int k)
  {
    int result = 0;
    while (n>0)
    {
      n = n-k;
      if(n>=0)
      {
        result = result + 1;
      }
    }
    return result;
  }

  // want to easily be able to view arrays
  static String showArray(int[] A)
  {
    StringBuilder builder = new StringBuilder();
    for(int i = 0;i<A.length;i++)
    {
      builder.append(A[i] + " ");
    }
    String result = builder.toString();
    return result;
  }


  static int[] genNewCombination(int[] potentialCombination,int[] maxOfEachStep)
  {
    if(potentialCombination.length == 0)
    {
      int[] tempArray = new int[maxOfEachStep.length];
      for (int i = 0;i<maxOfEachStep.length; i++)
      {
        tempArray[i] = 0;
      }
      return tempArray;
    }
    else
    {
      int i = 0;
      while(i<maxOfEachStep.length)
      {
        if(potentialCombination[i]<maxOfEachStep[i])
        {
        potentialCombination[i]=potentialCombination[i]+1;
        return potentialCombination;
        }
        else
        {
          potentialCombination[i] = 0;
          i = i + 1;
        }
      }
      return potentialCombination;
    }
  }

  static int testStepCombination(int[] potentialCombination, int[] stepList)
  {
    int result = 0;
    for(int i = 0;i<stepList.length;i++)
    {
      result = result + (potentialCombination[i] * stepList[i]);
    }
    return result;
  }

  static int combination(int n,int r)
  {
    int result = 1;
    if ((n-r)<r)
    {
      r = n-r;
    }
    for (int i = 0; i<r;i++)
    {
      result = result * (n-i);
      result = result / (i+1);
    }
    return result;
  }

  static int multCombination(int[] kBar)
  {
    int result = 1;
    int numerator = 0;
    for (int i = 0; i < kBar.length; i++)
    {
      numerator = numerator + kBar[i];
      result = result * combination(numerator, kBar[i]);
    }
    return result;
  }

}

File: Faster Than Recursive/StepFasterThanRecursive.py
# Step problem

def whilePositive(steps,k):
    n = 0
    while (steps>0):
        steps=steps-k
        if(steps>=0):
            n = n + 1
    return n

def isStepCombinationGood(steps,stepOptionList, itrTracker):
    if(sum([x*y for x,y in zip(itrTracker['lastCombination'],stepOptionList)])==steps):
        return True
    else:
        if(sum([x*y for x,y in zip(itrTracker['lastCombination'],stepOptionList)])==itrTracker['max']):
            itrTracker['done']=True
        return False


def genStepCombination(steps, stepOptionList, numPosSubs, itrTracker):
    combination = []
    if(len(itrTracker['lastCombination'])==0):
        for i in range(0,len(stepOptionList)):
            combination.append(0)
        itrTracker['lastCombination']=combination
        return isStepCombinationGood(steps,stepOptionList, itrTracker)

    increased = False
    i = 0
    while (itrTracker['done']==False & i<len(stepOptionList)):
        if(itrTracker['lastCombination'][i]<numPosSubs[i]):
            itrTracker['lastCombination'][i] = itrTracker['lastCombination'][i]+1
            increased = True
            return isStepCombinationGood(steps,stepOptionList, itrTracker)
        else:
            itrTracker['lastCombination'][i] = 0
            i = i + 1

def combination(n,r):
    if((n-r)<r):
        r=n-r
    result = 1
    for i in range(0,r):
        result=result*(n-i)
        result=result/(i+1)
    return result

def multCombination(kBar):
    result = 1
    numerator = 0
    for i in range(0,len(kBar)):
        numerator = numerator + kBar[i]
        result = result*combination(numerator,kBar[i])
    return result

def calcSteps(steps, stepOptionList):
    stepCombinationList = []
    numPosSubs = []
    nStepOption = len(stepOptionList)

    for i in range(0,nStepOption):
        numPosSubs.append(whilePositive(steps,stepOptionList[i]))

    itrTracker = {'nStepOptions': nStepOption,
     'lastCombination': [], 'max': sum([x*y for x,y in zip(stepOptionList,numPosSubs)]),  'done': False}
    while (itrTracker['done']==False):
        if(genStepCombination(steps, stepOptionList, numPosSubs, itrTracker)==True):
            stepCombinationList.append(list(itrTracker['lastCombination']))

    result = 0
    for i in range(0,len(stepCombinationList)):
        result = result + multCombination(stepCombinationList[i])
    return int(result)




stepOptionList = [1,2,3]
steps =40

print(calcSteps(steps,stepOptionList))
